File: Grind75/length_of_longest_substring.py
def length_of_longest_substring(s: str) -> int:
    last_seen = {}
    start = 0
    max_length = 0

    for end in range(len(s)):
        if s[end] in last_seen and last_seen[s[end]] >= start:
            start = last_seen[s[end]] + 1
        
        last_seen[s[end]] = end
        max_length = max(max_length, end - start + 1)
    
    return max_length

File: lengthOfLongestSubstring.js
function lengthOfLongestSubstring(s) {
  let max = 0;
  let l = 0;
  const lastSeen = new Map();

  for (let r = 0; r < s.length; r++) {
    const ch = s[r];

    if (lastSeen.has(ch) && lastSeen.get(ch) >= l) {
      l = lastSeen.get(ch) + 1;
    }

    max = Math.max(max, r - l + 1);
    lastSeen.set(ch, r);
  }

  return max;
}

// Time: O(n)
// Space: O(n)
